server/src/main/java/com/objectcomputing/checkins/services/memberprofile/retentionreport/RetentionReportServicesImpl.java

```diff
@@ -0,0 +1,352 @@
+package com.objectcomputing.checkins.services.memberprofile.retentionreport;
+
+import com.objectcomputing.checkins.services.memberprofile.MemberProfile;
+import com.objectcomputing.checkins.services.memberprofile.MemberProfileServices;
+
+import javax.inject.Singleton;
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+@Singleton
+public class RetentionReportServicesImpl implements RetentionReportServices {
+
+    private final MemberProfileServices memberProfileServices;
+
+    public RetentionReportServicesImpl(MemberProfileServices memberProfileServices) {
+        this.memberProfileServices = memberProfileServices;
+    }
+
+    @Override
+    public RetentionReportResponseDTO report(RetentionReportDTO request) {
+        RetentionReportResponseDTO response;
+        List<MemberProfile> memberProfiles = memberProfileServices.findAll();
+        LocalDate periodStartDate = getIntervalStartDate(request.getStartDate(), request.getFrequency());
+        LocalDate periodEndDate = getIntervalEndDate(request.getEndDate(), request.getFrequency());
+        if (request.getFrequency() == FrequencyType.DAILY) {
+            response = getReportValues(periodStartDate, periodEndDate, memberProfiles, FrequencyType.DAILY);
+        } else if (request.getFrequency() == FrequencyType.WEEKLY) {
+            response = getReportValues(periodStartDate, periodEndDate, memberProfiles, FrequencyType.WEEKLY);
+        } else {
+            response = getReportValues(periodStartDate, periodEndDate, memberProfiles, FrequencyType.MONTHLY);
+        }
+
+        return response;
+    }
+
+    private LocalDate getIntervalStartDate(LocalDate periodStartDate, FrequencyType frequency) {
+        LocalDate startDate;
+        if (frequency == FrequencyType.DAILY) {
+            startDate = periodStartDate;
+        } else if (frequency == FrequencyType.WEEKLY) {
+            DayOfWeek day = periodStartDate.getDayOfWeek();
+            switch (day) {
+                case SUNDAY:
+                    startDate = periodStartDate.plusDays(6);
+                    break;
+                case MONDAY:
+                    startDate = periodStartDate.plusDays(5);
+                    break;
+                case TUESDAY:
+                    startDate = periodStartDate.plusDays(4);
+                    break;
+                case WEDNESDAY:
+                    startDate = periodStartDate.plusDays(3);
+                    break;
+                case THURSDAY:
+                    startDate = periodStartDate.plusDays(2);
+                    break;
+                case FRIDAY:
+                    startDate = periodStartDate.plusDays(1);
+                    break;
+                default:

```

Comment:
Do you not need a SATURDAY in here? Or is the periodStartDate Sat by default?

Reply:
Right, the default is Saturday, the other cases are adjusting to the Saturday of the week.